src/Adapter/TaxRulesGroup/AbstractTaxRulesGroupHandler.php

```diff
@@ -0,0 +1,249 @@
+            throw new TaxRuleNotFoundException(
+                sprintf('Tax rule with id "%s" was not found.', $taxRuleIdValue)
+            );
+        }
+
+        return $taxRule;
+    }
+
+    /**
+     * Deletes legacy TaxRule
+     *
+     * @param TaxRule $taxRule
+     *
+     * @return bool
+     *
+     * @throws CannotDeleteTaxRuleException
+     */
+    protected function deleteTaxRule(TaxRule $taxRule): bool
+    {
+        try {
+            return $taxRule->delete();
+        } catch (PrestaShopException $e) {
+            throw new CannotDeleteTaxRuleException(
+                sprintf('An error occurred when deleting tax rule object with id "%s".', $taxRule->id)
+            );
+        }
+    }
+
+    /**
+     * @param TaxRulesGroup $taxRulesGroup
+     *
+     * @throws TaxRulesGroupConstraintException
+     * @throws PrestaShopException
+     */
+    protected function validateTaxRulesGroupFields(TaxRulesGroup $taxRulesGroup): void
+    {
+        if (!$taxRulesGroup->validateFields(false)) {
+            throw new TaxRulesGroupConstraintException(
+                'Tax rules group contains invalid field values'
+            );
+        }
+    }
+
+    /**
+     * @param TaxRule $taxRule
+     *
+     * @throws PrestaShopException
+     * @throws TaxRuleConstraintException
+     */
+    protected function validateTaxRuleFields(TaxRule $taxRule): void
+    {
+        if (!$taxRule->validateFields(false)) {
+            throw new TaxRuleConstraintException(
+                'Tax rule contains invalid field values'
+            );
+        }
+    }
+
+    /**
+     * @param int $taxRulesGroupId
+     * @param int $countryId
+     * @param int $stateId
+     * @param int|null $taxRuleId
+     *
+     * @return bool
+     */
+    protected function assertUniqueBehaviorTaxRuleForCountry(
+        int $taxRulesGroupId,
+        int $countryId,
+        int $stateId,
+        ?int $taxRuleId = null
+    ): bool {
+        $errors = $this->validator->validate([
+            'taxRulesGroupId' => $taxRulesGroupId,
+            'country' => $countryId,
+            'state' => $stateId,
+            'taxRuleId' => $taxRuleId,
+        ], new UniqueTaxRuleBehavior());
+
+        return 0 !== count($errors);
+    }
+
+    /**
+     * @param string $zipCode
+     * @param int $countryId
+     *
+     * @return bool
+     */
+    protected function assertIsValidZipCode(
+        string $zipCode,
+        int $countryId
+    ): bool {
+        $errors = $this->validator->validate([
+            'zipCode' => $zipCode,
+            'country' => $countryId,
+        ], new ZipCodeRange());
+
+        return 0 !== count($errors);
+    }
+
+    /**
+     * @param CountryDataProvider $countryDataProvider
+     * @param int $langId
+     * @param CountryId|null $countryId
+     *
+     * @return int[]
+     *
+     * @throws CountryNotFoundException
+     */
+    protected function getCountryForTaxRule(
```

Comment:
```suggestion
    protected function getCountryIDsForTaxRule(
```
This function seems to return country IDs. Moreover it does not seem to find country IDs for a tax rule, but rather for a `$langID` and/or `$countryID` 🤔

Reply:
This function does 2 things. If `$countryId` is provided, it returns `array($countryId)`. Else, it finds the country IDs for this lang ID.
=> we need this function to do only 1 thing 😉 

I suggest to move the
```
if ($countryId !== null && $countryId->getValue() > 0) {
    $selectedCountries[] = $countryId->getValue();
}
```
logic outside of this function

Reply:
If you do this, this function becomes very simple. It's a call to `$countryDataProvider->getCountries($langId);` followed by the `array_map()` 😄